Changes from GOSL reporting

paddytobias opened the pull request.

paddytobias committed:
Adding feature for when a single shadow is provided amongst two or more shadows on other columns in order to fix merge issue

paddytobias committed:
Increment version number

paddytobias committed:
Adding news

paddytobias committed:
updates to docs

paddytobias commented:
Hey @kinto-b , just making some changes to support GOSL reporting. Would you mind reviewing when you get a chance?

kinto-b commented:
@paddytobias  Yo! 

Is the intention here to use the name provided in the shadow (for one-place shadows) as the name of the column? For example `prj_project(blah, list(v=c(n="{N}"))` would rename `v` to `n` in the projected df?

The only issue I can see is that providing the same named one-place shadow for multiple columns will now throw an error. For instance:

```r
tbl <- mtcars %>% 
  prj_tbl_rows(
    Cylinders = cyl,
    Transmission = list(Automatic = am %in% 0, Manual = am %in% 1),
    ) %>% 
  prj_tbl_cols(
    `V-Shaped` = col_freq(n = vs %in% 1, N = vs %in% 0:1),
    `Not V-shaped` = col_freq(n = vs %in% 0, N = vs %in% 0:1)
    ) %>% 
  prj_tbl_summarise()
  
tbl %>% 
  prj_project(list(
    `V-Shaped` = c(p = "{signif(p, 2)} ({n})"),
    `Not V-shaped` = c(p = "{signif(p, 2)} ({n})")
  ))
```

It would be good to handle this behind the scenes and signal a warning.

Other than that, could you also add:
* a test case or two, and
* an example into the NEWS and the function `@examples` section to illustrate the new behaviour?

paddytobias committed:
Changes:
* Adding renaming of dupe column names as suggested in #29
* Adding examples to NEWS and function documentation
* Adding tests

paddytobias committed:
Starting to move functions from srcreporting into projectable

paddytobias committed:
Adding tests for dupe column names, and preserving formatting

paddytobias committed:
Adding to examples for dupe column names

paddytobias committed:
Resolving test error

paddytobias commented:
Thanks @kinto-b for thinking this through more, I've added some internal handling for duplicate column names. A increment suffix is added to the name and a warning is thrown

``` r
library(projectable)
library(magrittr)
#> Warning: package 'magrittr' was built under R version 4.0.5
tbl <- mtcars %>% 
  prj_tbl_rows(
    Cylinders = cyl,
    Transmission = list(Automatic = am %in% 0, Manual = am %in% 1),
  ) %>% 
  prj_tbl_cols(
    `V-Shaped` = col_freq(n = vs %in% 1, N = vs %in% 0:1),
    `Not V-shaped` = col_freq(n = vs %in% 0, N = vs %in% 0:1)
  ) %>% 
  prj_tbl_summarise()

tbl %>% 
  prj_project(list(
    `V-Shaped` = c(p = "{signif(p, 2)} ({n})"),
    `Not V-shaped` = c(p = "{signif(p, 2)} ({n})")
  ))
#> Warning in prj_cast_shadow(.data, .digits = .digits): 
#> Column name duplicated: `p`
#> Resolving by incrementing...
#> # A tibble: 5 x 4
#>   row_spanner  rows      p1        p2       
#> * <col_row>    <col_row> <glue>    <glue>   
#> 1 Cylinders    4         0.91 (10) 0.09 (1) 
#> 2 Cylinders    6         0.57 (4)  0.43 (3) 
#> 3 Cylinders    8         0 (0)     1 (14)   
#> 4 Transmission Automatic 0.37 (7)  0.63 (12)
#> 5 Transmission Manual    0.54 (7)  0.46 (6)
```

<sup>Created on 2022-05-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>


Have also added to the tests and added examples to the  NEWS and function documentation

paddytobias committed:
Adding util from screporting, needed for the public functions

paddytobias committed:
Adding dplyr as an import

paddytobias committed:
Debugging

paddytobias committed:
Adding myself to authors :)

paddytobias committed:
Updating docs

kinto-b commented on R/public.R:
```diff
@@ -0,0 +1,102 @@
+public_num <- function(x, N, digits = 1, min_N = 25) {
+  N <- match_lengths(N, x)
+
+  dplyr::case_when(
+    is.na(x) | is.na(N) | N %in% 0 ~ "", # No records in cell
+    N < min_N ~ "n/a", # Too few records in cell
+    TRUE ~ dec_dig3(x, digits)
+  )
+}
```
I don't think importing the `dplyr` dep is worth it for just this case_when and the one below. This can easily be changed to a series of assignments:

```r
x[is.na(x) | is.na(N) | N %in% 0] <- ""
# etc.
```

kinto-b commented on R/public.R:
```diff
@@ -0,0 +1,102 @@
+
+#' @export
+#' @rdname public_formats
+public_num <- function(x, N, digits = 1, min_N = 25) {
```
I'm still not entirely convinced these belong here. They're much more related to how SRC does reporting than the `projectable` class and associated methods. Also would be a bit cleaner to introduce these in a separate PR so that they have their own commit

kinto-b reviewed (commented):
Thanks Paddy :) Looking good! Just a small few comments

kinto-b commented on R/prj_project.R:
```diff
@@ -106,19 +136,41 @@ prj_cast_shadow <- function(.data, .digits = NULL) {
 
+  # Renaming duplicated column names
+  if (any(duplicated(names(out)))) {
+    warning("\nColumn name duplicated: `", names(out)[duplicated(names(out))], "`\nResolving by incrementing...")
```
Oh actually one other thing, could you add `call.=FALSE` to `warning()` here?

kinto-b commented on R/prj_project.R:
```diff
@@ -106,19 +136,41 @@ prj_cast_shadow <- function(.data, .digits = NULL) {
+  if (any(duplicated(names(out)))) {
+    warning("\nColumn name duplicated: `", names(out)[duplicated(names(out))], "`\nResolving by incrementing...")
+
+    which_dups <- which(names(out) == names(out)[duplicated(names(out))])
+    rename_dups <- paste0(names(out)[which(names(out) == names(out)[duplicated(names(out))])],
+                          seq_len(length(which(names(out) == names(out)[duplicated(names(out))]))))
+
+
+    col_labels[which_dups] <- rename_dups
+
+    mapply(function(elem, i) {
+      names(out)[elem] <<- rename_dups[i]
+    }, which_dups, seq_len(length(which_dups)))
+
+  }
+
```
Also, it's probably a bit more readable to just use

```r
 if (any(duplicated(names(out)))) {
  warning(...)
  names(out) <- make.unique(names(out))
}
```

justanna95 committed:
removing public presenting functions from projectable to be in srcreporting #29

justanna95 committed:
remove dependancy on dplyr - only reference was in public.R which was moved to srcreporting anyway #29

justanna95 committed:
formatting clean up

justanna95 committed:
update versioning and build changes

justanna95 commented:
Hey @paddytobias I've moved the public functions back to srcreporting and updated the versioning etc. I don't seem to have permissions to push my commits to github though.

kinto-b commented:
Hey @justanna95, I've just given you permissions, so should be good to go :)
